### PR TITLE
Check UFL FEniCS merge

### PR DESCRIPTION
**Pull request by nbouziani**
This checks that we are not breaking anything by merging UFL FEniCS. See the corresponding [UFL PR](https://github.com/firedrakeproject/ufl/pull/32).

This PR will also contain the fixes resulting from recent UFL changes since some deprecated UFL objects got removed (e.g. `ufl.FacetElement`).

This PR is required for merging the UFL PR and therefore merging the recent dual space changes.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -42,7 +42,7 @@ jobs:
       - name: Build Firedrake
         run: |
           cd ..
-          ./firedrake/scripts/firedrake-install $COMPLEX --venv-name build --tinyasm --disable-ssh --minimal-petsc --slepc --documentation-dependencies --install thetis --install gusto --install icepack --install irksome --install femlium --no-package-manager || (cat firedrake-install.log && /bin/false)
+          ./firedrake/scripts/firedrake-install $COMPLEX --venv-name build --tinyasm --disable-ssh --minimal-petsc --slepc --documentation-dependencies --install thetis --install gusto --install icepack --install irksome --install femlium --no-package-manager --package_branch ufl merge_fenics || (cat firedrake-install.log && /bin/false)
```
Comment (ReubenHill) [inline .github/workflows/build.yml:45]: This is failing because you've written `--package_branch` instead of `--package-branch` (`-` rather than `_`)

Comment (nbouziani) [reply .github/workflows/build.yml]: Yeah I was expecting tests to fail because master is broken so didn't check. Thanks!

**Comment by nbouziani** [comment]
There is still a test using deprecated elements (`FacetElement` and `InteriorElement`), namely: [test_cg_poisson.py](https://github.com/firedrakeproject/firedrake/blob/master/tests/slate/test_cg_poisson.py)

This test checks CG performance, I am not sure if it is inherently dependent on these elements, in which case I guess we should remove it, or if we should reformulate it.

**Comment by ReubenHill** [comment]
Why is this diff so big? Are you attempting to deal with upstream UFL changes as well as testing the recent merge of dual?

**Comment by nbouziani** [comment]
> Why is this diff so big? Are you attempting to deal with upstream UFL changes as well as testing the recent merge of dual?

No, this is about upstream UFL changes before dual that didn't get merged and are breaking Firedrake.

File: tests/regression/test_interior_elements.py [hunk 2]
```diff
@@ -1,14 +0,0 @@
...
-    mesh = UnitSquareMesh(5, 5)
-    V_elt = FiniteElement("RT", triangle, 2)
-    W2_elt = InteriorElement(V_elt)
-    W2 = FunctionSpace(mesh, W2_elt)
-    g = Function(W2).assign(1)
-    n = FacetNormal(mesh)
-
-    # For H(div) elements, interior dofs have u.n = 0 on facets
-    assert abs(assemble(dot(g('+'), n('+'))*dS)) < 1e-14
-    assert abs(assemble(dot(g, n)*ds)) < 1e-14
```
Comment (wence-) [inline tests/regression/test_interior_elements.py:14]: These tests should be ported to use `RestrictedElement(..., restriction="interior")` I think

File: tests/regression/test_facet_elements.py [hunk 3]
```diff
@@ -1,18 +0,0 @@
...
-    V = FunctionSpace(mesh, V_elt)
-    W1 = FunctionSpace(mesh, W1_elt)
-    f = Function(V).assign(1)
-    g = Function(W1).assign(1)
-    n = FacetNormal(mesh)
-
-    assert abs(assemble(inner(f, f)*dx) - assemble(inner(g, g)*dx)) < 2e-12
-    assert abs(assemble(inner(f, f)*ds) - assemble(inner(g, g)*ds)) < 2e-12
-    assert abs(assemble(inner(f, n)*ds) - assemble(inner(g, n)*ds)) < 2e-12
-    assert abs(assemble(inner(f('+'), n('+'))*dS) - assemble(inner(g('+'), n('+'))*dS)) < 2e-12
```
Comment (wence-) [inline tests/regression/test_facet_elements.py:18]: This should be tested using `RestrictedElement(..., restriction="facet")` ?

**Comment by nbouziani** [comment]
> > No, this is about upstream UFL changes before dual that didn't get merged and are breaking Firedrake.
> 
> Aha! So hopefully you've fixed #2595 ?

If you are referring to what got merged in `firedrakeproject/ufl` but not in `FEniCS/ufl`, that's a different issue for which I think the solution is to update and merge [#107](https://github.com/FEniCS/ufl/pull/107/files). This PR is the other way around.

File: tests/regression/test_interior_elements.py (hunk 2 above)
Comment (nbouziani) [reply tests/regression/test_interior_elements.py:14]: Oh yes, thanks!

File: tests/regression/test_facet_elements.py (hunk 3 above)
Comment (nbouziani) [reply tests/regression/test_facet_elements.py]: Done thanks!

**Comment by ReubenHill** [comment]
> > > No, this is about upstream UFL changes before dual that didn't get merged and are breaking Firedrake.
> > 
> > 
> > Aha! So hopefully you've fixed #2595 ?
> 
> If you are referring to what got merged in `firedrakeproject/ufl` but not in `FEniCS/ufl`, that's a different issue for which I think the solution is to update and merge [#107](https://github.com/FEniCS/ufl/pull/107/files). This PR is the other way around.

I don't think that was the actual cause of the issue, instead I believe that that merge having happened stopped us updating UFL from upstream for a while (because we had diverged) and then, when we did, we had regressed as outlined in #2595 . The suggested fix is at the bottom of that issue and is not related to trimmed serendipity ( [#107](https://github.com/FEniCS/ufl/pull/107/files)).

**Comment by nbouziani** [comment]
> > > > No, this is about upstream UFL changes before dual that didn't get merged and are breaking Firedrake.
> > > 
> > > 
> > > Aha! So hopefully you've fixed #2595 ?
> > 
> > 
> > If you are referring to what got merged in `firedrakeproject/ufl` but not in `FEniCS/ufl`, that's a different issue for which I think the solution is to update and merge [#107](https://github.com/FEniCS/ufl/pull/107/files). This PR is the other way around.
> 
> I don't think that was the actual cause of the issue, instead I believe that that merge having happened stopped us updating UFL from upstream for a while (because we had diverged) and then, when we did, we had regressed as outlined in #2595 . The suggested fix is at the bottom of that issue and is not related to trimmed serendipity ( [#107](https://github.com/FEniCS/ufl/pull/107/files)).

This PR is about merging UFL upstream changes until dual while the issue you mentioned is about something that already got merged and that is a bug in both UFLs at the moment. Therefore, it's a different problem as one would fix issue [#2595](https://github.com/firedrakeproject/firedrake/issues/2595) by making a PR against UFL upstream that will then get merged and only show up on our end once we have merged what happened in-between, namely dual and the content of this PR which is breaking Firedrake. We can discuss this in today's meeting.

**Review by dham — APPROVED** [review]
I think this looks fine. Please remove the build.yml change and we can get merging.

**Pull request merged**